restore-sql-server-transaction-logs/Function/main.py

```diff
@@ -0,0 +1,274 @@
+    if object_name.endswith('/'):
+        logger.info("Folder creation encountererd. Operation skipped.")
+        return "Operation skipped."
+
+    valid_env_v, msg, processed_bucket_name, use_fixed_file_name_format, max_request_attempts, max_request_fetch_time_seconds, max_operation_fetch_time_seconds = get_environment_variables()
+
+    if not(valid_env_v):        
+        raise RuntimeError(f"Operation failed. Environment variables validation failed: {msg}.")        
+    
+    if (use_fixed_file_name_format.lower()=="true"):        
+        logger.info("Using fixed file name format")
+        valid_file_name, msg, csql_instance_name, database_name, backup_type, no_recovery = extract_info_from_file_name(object_name)
+
+        if not(valid_file_name):
+            raise RuntimeError(f"Operation failed. File Name validation failed: {msg}.")
+    else:
+        logger.info(f"Using metadata")
+        valid_metadata, msg, csql_instance_name, database_name, backup_type, no_recovery = extract_info_from_file_metadata(source_bucket_name, object_name)        
+
+        if not(valid_metadata):
+            raise RuntimeError(f"Operation failed. Metadta validation failed: {msg}.")
+
+    request_attempts = 0
+    log_file_processed = False
+
+    creds, project_id = google.auth.default()
+    auth_req = google.auth.transport.requests.Request()
+    creds.refresh(auth_req)
+
+    service = googleapiclient.discovery.build('sqladmin', 'v1beta4', credentials=creds)
+
+    storage_client = storage.Client()
+    source_bucket = storage_client.bucket(source_bucket_name)    
+
+    import_context_body = get_import_context(source_bucket_name,object_name,database_name,backup_type, no_recovery)
+
+    while log_file_processed == False:
+
+        request_attempts += 1
+
+        try:
+            
+            logger.info(f"Attempt to process object {object_name}. Try {request_attempts} out of {max_request_attempts}...")
+
+            request = service.instances().import_(project=project_id, instance=csql_instance_name, body=import_context_body)
+            response = request.execute()
+            
+            logger.info(f"Executed request: {import_context_body}")
+                        
+            operation_status_done = False
+            operation_status_fetch_attempts=0
+
+            while not operation_status_done:
+
+                operation_status_fetch_attempts += 1
+                operation_sleep_time_seconds = min(max_operation_fetch_time_seconds, 5*operation_status_fetch_attempts)
+                
+                logger.info(f"Waiting for {operation_sleep_time_seconds} seconds to check for the operation status...")
+                
+                sleep(operation_sleep_time_seconds)
+                
+                resp = service.operations().get(project=project_id, operation=response["name"]).execute()
+
+                logger.info(f"Fetched operation status: {resp['status']}")
+
+                if resp["status"] == "DONE":

```

Comment:
If anything else happens besides `DONE`, we get no feedback, right? This should at least log an error or something.

Reply:
There is the inner loop that does the following:

- Increment the operation fetch status attempts;
- Wait some seconds;
- Get the status and write it to the log.
- If it’s “DONE”, inspect if there were any errors.

I would say this is enough because, there is a max fetch status attempts defined and we do log the status each time we ask for it.

Reply:
Now relying on the retry mechanism of the function/eventarc, but this loop remans as we are pooling the status of the operation periodically.

Reply:
LGTM